drivers: STM32 dualcore concurrent register access protection with HSEM

ABOSTM opened the pull request:
In case of dualcore, STM32H7, STM32W and STM32MP1,
protect concurrent register write access with HSEM.
Done for following drivers:
clock_control, counter, flash, gpio, interrupt_controller

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -218,6 +241,12 @@ static int stm32_clock_control_init(struct device *dev)
 
 #ifdef CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
 
+#if defined(CONFIG_STM32H7_DUAL_CORE)
```
I wonder if hsem protection is required here since only M7 run this function (and only once at boot for now)

ABOSTM replied:
It is true that M4 doesn't run this function, but it is still able to access  RCC registers.
And even if today it should ot happends while M7 is configuring clocks,
I prefer to keep it in case it becomes necessary (with introduction of power management for example)

gmarull commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -64,7 +65,15 @@ static int rtc_stm32_start(struct device *dev)
 {
 	ARG_UNUSED(dev);
 
+#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE) || defined(CONFIG_SOC_SERIES_STM32MP1X)
```
I think it would be better in general to have a define on top like

```c
#define HAS_HSEM (defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE) || defined(CONFIG_SOC_SERIES_STM32MP1X))
```
 and use:

```c
#if HAS_HSEM
// code for SoC that have HSEM
#endif
```

It makes code more readable and code requires less changes when adding support for new SoC.

ABOSTM replied:
Fully reworked. Done

gmarull commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,64 @@
+ 
+#endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
+
+#define HSEM_TIMEOUT   0xFFFF
```
value is not documented

ABOSTM replied:
Done

gmarull commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,64 @@
+/* Index of the semaphore used to access GPIO */
+#define CFG_HW_GPIO_SEMID                                       0
+
+/* Index of the semaphore used to access the EXTI */
```
`/**` to make it *doxygenable*?

ABOSTM replied:
Done

ABOSTM commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -174,6 +175,11 @@ int gpio_stm32_configure(u32_t *base_addr, int pin, int conf, int altf)
+	uint32_t timeout = HSEM_TIMEOUT;
+	while (LL_HSEM_1StepLock(HSEM, CFG_HW_GPIO_SEMID) && (--timeout != 0)) {
+	}
+#endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE || CONFIG_SOC_SERIES_STM32MP1X)*/
```
Done

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -9,6 +9,7 @@
 #include <drivers/clock_control.h>
 #include <sys/util.h>
 #include <drivers/clock_control/stm32_clock_control.h>
+#include <../soc/arm/st_stm32/common/stm32_hsem.h>
```
Done

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -9,6 +9,7 @@
 #include <drivers/clock_control.h>
 #include <sys/util.h>
 #include <drivers/clock_control/stm32_clock_control.h>
+#include <../soc/arm/st_stm32/common/stm32_hsem.h>
```
Path was correct, but whaterver it has been modified.

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -47,12 +48,20 @@
 {
+#if defined(CONFIG_STM32H7_DUAL_CORE)
+	uint32_t timeout = HSEM_TIMEOUT;
+	while (LL_HSEM_1StepLock(HSEM, CFG_HW_RCC_SEMID) && (--timeout != 0)) {
```
This is the way it was done in STM32Cube Example. Nevertheless, I agree I can to use Zephyr error handling in case of timeout. Done

gmarull commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,111 @@
+{
+#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE) \
+    || defined(CONFIG_SOC_SERIES_STM32MP1X)
+	u32_t timeout = HSEM_TIMEOUT;
```
I would actually change the name to something like `HSEM_LOCK_RETRIES`, as the variable/definition do not represent time.

erwango commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -1,5 +1,10 @@
 # SPDX-License-Identifier: Apache-2.0
 
+zephyr_include_directories_ifdef(
+	CONFIG_SOC_FAMILY_STM32
+	${ZEPHYR_BASE}/soc/arm/st_stm32/common
+	)
+
```
There should be a way to get rid of this.

erwango replied:
This is actually not needed. The fact that you modified the #include directive was sufficient.
You can remove it.

ABOSTM replied:
Replaced by adding include directive in soc/arm/st_stm32/CmakeLists.txt

erwango commented on soc/arm/st_stm32/stm32h7/soc.h:
```diff
@@ -18,13 +18,8 @@
-#define LL_HSEM_MASK_1 (1 << LL_HSEM_ID_1)
-
 #include <stm32h7xx_ll_hsem.h>
+#include "../common/stm32_hsem.h"
```
#include "stm32_hsem.h"

ABOSTM replied:
Done

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -10,6 +10,8 @@
 #include <sys/util.h>
 #include <drivers/clock_control/stm32_clock_control.h>
 
+#include "stm32_hsem.h"
```
This one is not technically needed since included in soc.h, which is included here.

ABOSTM replied:
Done

erwango commented on soc/arm/st_stm32/stm32mp1/soc.h:
```diff
@@ -26,6 +26,8 @@
  * core SOC-specific headers.
  */
 #include <kernel_includes.h>
+#include <stm32mp1xx_ll_hsem.h>
```
Break a line before this include

ABOSTM replied:
Done

erwango commented on soc/arm/st_stm32/stm32wb/soc.h:
```diff
@@ -25,7 +25,7 @@
 /* Add include for DTS generated information */
 #include <devicetree.h>
-
+#include <stm32wbxx_ll_hsem.h>
```
Is that required ?

erwango replied:
Please discard this one!

ABOSTM replied:
It is required: if I remove it, I face link issue when building wb.

erwango commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,111 @@
+/**
+ * @brief Lock Hardware Semaphore
+ */
+static inline void stm32_hsem_lock(u32_t  hsem)
```
Let's rename as z_stm32_hsem_lock to make clear this is a zephyr dedicated function

ABOSTM replied:
Done

erwango commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,111 @@
+/**
+ * @brief Release Hardware Semaphore
+ */
+static inline void stm32_hsem_release(u32_t  hsem)
```
z_stm32_hsem_release

ABOSTM replied:
Done

pabigot commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -147,6 +152,7 @@ static inline int stm32_clock_control_off(struct device *dev,
 	default:
 		return -ENOTSUP;
```
This needs to be something like:
```
    rc = -ENOTSOP;;
   goto out;
```
otherwise you're returning while `hsem_lock` is still held.

Please check for this problem elsewhere too.

ABOSTM replied:
Nice catch, thanks. Done

pabigot commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -269,6 +277,8 @@ static int stm32_clock_control_init(struct device *dev)
 	#error "CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL not selected"
 #endif /* CLOCK_STM32_SYSCLK_SRC_PLL */
 
+	z_stm32_hsem_release(CFG_HW_RCC_SEMID);
```
In line 269 `config_bus_prescalers()`, which takes the lock, is being invoked while the lock is held.  Is that correct (i.e. does this lock count)?

ABOSTM replied:
This is not a problem because we use 1step lock mechanism which lock again CoreID only (compare current coreID with the one which may have lock the semaphore).

pabigot replied:
So the `z_stm32_hsem_release()` invoked at the end of `config_bus_prescalers()` will also be ignored?  Or will the lock have been released, so the `z_stm32_hsem_release()` called here will have no effect?

ABOSTM replied:
Humm ... right, semaphore will be release at first occurence of `z_stm32_hsem_release(`), which is not what is expected.
I get ride of `config_bus_prescalers()`  which was excusively used here, and thus only 1 Lock/release for the whole init processing.
Thanks for your accurate review.

pabigot reviewed (changes requested):
Multiple functions have an exit path while the lock is still held.  These should be fixed.

Also it appears that there are nested lock calls; I don't know the hardware to say whether this is safe.

arnopo commented on drivers/clock_control/clock_stm32_ll_mp1.c:
```diff
@@ -19,6 +21,8 @@ static inline int stm32_clock_control_on(struct device *dev,
 	ARG_UNUSED(dev);
 
+	z_stm32_hsem_lock(CFG_HW_RCC_SEMID);
+
```
There are 2 register banks to enable/disable the clocks, one for the MCU, one for the MPU so no HWSEM needed in this file

ABOSTM replied:
Done

arnopo commented on soc/arm/st_stm32/common/stm32_hsem.h:
```diff
@@ -0,0 +1,110 @@
+
+/** Index of the semaphore used to access the RNG */
+#define CFG_HW_RNG_SEMID                    2
+
```
please don't define semaphore 2 to 5 , this would means that the zephyr could access to these IPs, but this is not the case as reserved for the MPU usage. This could lead to corruption

pabigot commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -100,8 +85,10 @@ static inline int stm32_clock_control_on(struct device *dev,
 		LL_APB4_GRP1_EnableClock(pclken->enr);
 		break;
 	default:
+		z_stm32_hsem_release(CFG_HW_RCC_SEMID);
```
I won't block for this but a solution idiom of:

```
int rc = 0;

z_stm32_hsem_lock(CFG_HW_RCC_SEMID);
switch (pclken->bus) {
case STM32_CLOCK_BUS_AHB1:
        LL_AHB1_GRP1_EnableClock(pclken->enr);
        break;
...
default:
	rc = -ENOTSUP;
	break;
}

z_stm32_hsem_release(CFG_HW_RCC_SEMID);

return rc;
```

would be easier to understand because there's clearly one unlock for each lock.   It's also more consistent with future style guidance to have one return statement per function.

ABOSTM replied:
Done

pabigot commented on drivers/flash/flash_stm32.c:
```diff
@@ -65,11 +64,7 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
-#endif /* CONFIG_SOC_SERIES_STM32WBX */
+	z_stm32_hsem_lock(CFG_HW_FLASH_SEMID);
 
 	k_sem_take(&FLASH_STM32_PRIV(dev)->sem, K_FOREVER);
```
This seems wrong for a couple reasons.

First, the order is wrong.  `k_sem_take()` can put the calling thread to sleep for as long as it takes a previous invocation to complete.  Presumably that invocation has already taken the HSEM lock, so again we have a double-lock / premature unlock situation (the lock taken here will have been released by the unlock that releases the semaphore this thread is waiting for).

So at a minimum HSEM must be locked and unlocked while the thread holds the lock on the device.

Second, this means HSEM is kept locked for as long as a flash operation is in progress, which is a very long time.  It seems likely other things might invoke operations that also take and release the HSEM lock.

I suspect this driver needs to use much finer-grained locking of HSEM, just as most of the other drivers do.

I'm also concerned now that if any of the API that takes and release HSEM can be invoked from an interrupt (e.g. `gpio_pin_set()`) that the locking regiment will be broken because the nested lock operation has no effect, but the unlock will release both locks.  Not only here, but possibly everywhere else that this new API is used to lock and unlock while interrupts are enabled.

I think this PR needs an architecture overview from an STM maintainer.  It may require a different solution.

ABOSTM replied:
First: Reverse order -> done.
Second: Please note that there are several distinct HSEM (cuurently 1 for Flash, 1for GPIO, 1 for EXTI, ...). So we have some granularity.
And if 

> other things might invoke operations that also take and release the HSEM lock

it must be on the same Flash HSEM. So it is no so likekely. Nevertheless, it is still possible to have both cores trying to write flash, so, specially for FLASH use case which may hold the lock for long time,  I allow to wait forever.

Concerning interrupt use case, I agree on this issue, but it is not specific to HSEM or dualcore. 
As you know there are lot of discussions around multithreading and/or interruptions (#18970 , #6184, #24511,  ...).
My view for HSEM is that: first in general we should tend to have a single driver able to access register of a specific IP (example only clock control should access to RCC registers, ...), then if this driver is protected against multithread and interruptions, then HSEM will implicitly be protect too.
But to be honest I didn't clearly understand how drivers will be protected against Interruptions.

For the specific case of gpio: `gpio_pin_set()` doesn't make use of HSEM, because it uses specific register able to address a single bit of a register. Nevertheless I noticed that it is not the case for all functions (like toggle). I will investigate to see how far I can get ride of HSEM and use this gpio specific register (in separate PR).

pabigot reviewed (dismissed):
One clear problem newly identified, and an overall impression that the solution needs to be reconsidered to take into account protecting the lock status from mutation by interrupts.  The latter is outside my area of expertise.

ABOSTM commented:
I made a rebase of this PR.
Is there further comments?
or can it be reviewed and merged ?

ABOSTM committed:
drivers: STM32 dualcore concurrent register access protection with HSEM

In case of dualcore, STM32H7, STM32W and STM32MP1,
protect concurrent register write access with HSEM.
Done for following drivers:
clock_control, counter, flash, gpio, interrupt_controller

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

pabigot commented on soc/arm/st_stm32/stm32mp1/soc.h:
```diff
@@ -24,6 +24,9 @@
 #include <devicetree.h>
 
+#include <stm32mp1xx_ll_hsem.h>
+
```
extra newline?

ABOSTM replied:
Done

pabigot reviewed (commented):
Just commentary: I don't have expertise to judge the technical merit of this SOC-specific solution.

I've removed my NAK under the assumption that the mis-ordered lock sequences have been fixed, and each lock is paired with an unlock (no early exits/inappropriate nested locking).  From a cursory look only at the diffs github provides this seems to be true, but I have not looked at the unmodified areas of the code to confirm it.

I find it confusing that `z_stm32_hsem_lock` is paired with `z_stm32_hsem_release`.  Pairing `lock` with `unlock`, or `request` with `release`, would make be more consistent with existing usage (request/release only if the lock is recursive, but I think it isn't).  Since this is STM32-specific API I don't insist on this.

ABOSTM commented:
I replaced `z_stm32_hsem_release() `by `z_stm32_hsem_unlock()`.
I also fixed rebase regression (include model).

ABOSTM commented:
Sanitychek failure: ' Sanitycheck / qemu_arc_hs:tests/portability/cmsis_rtos_v1/portability.cmsis_rtos_v1 / tests/portability/cmsis_rtos_v1/portability.cmsis_rtos_v1'

arnopo reviewed (approved):
just a minor comment, else LGTM

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -271,6 +274,8 @@ static int stm32_clock_control_init(struct device *dev)
 	#error "CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL not selected"
 #endif /* CLOCK_STM32_SYSCLK_SRC_PLL */
 
+	z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
```
Done
